metastore/src/java/org/apache/hadoop/hive/metastore/HiveProtoEventsCleanerTask.java

```diff
@@ -0,0 +1,176 @@
+    LocalDate expiredDate
+            = LocalDateTime.ofEpochSecond((clock.getTime() - ttl) / 1000, 0, ZoneOffset.UTC).toLocalDate();
+    return "date=" + DateTimeFormatter.ISO_LOCAL_DATE.format(expiredDate);
+  }
+
+  /**
+   * Path filters to include only expired date partitions based on TTL.
+   */
+  private static final PathFilter expiredDatePartitionsFilter = new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      String dirName = path.getName();
+      return ((dirName.startsWith("date="))
+              && (dirName.compareTo(expiredDatePtn) <= 0));
+    }
+  };
+
+  /**
+   * Finds the expired date partitioned events directory based on TTL and delete them.
+   */
+  private void cleanupDir(Path eventsBasePath) {
+    LOG.debug("Trying to delete expired proto events from " + eventsBasePath);
+    try {
+      FileSystem fs = FileSystem.get(eventsBasePath.toUri(), conf);
+      if (!fs.exists(eventsBasePath)) {
+        return;
+      }
+      FileStatus[] statuses = fs.listStatus(eventsBasePath, expiredDatePartitionsFilter);
+      for (FileStatus dir : statuses) {
+        deleteDir(fs, dir);
+        LOG.info("Deleted expired proto events dir: " + dir.getPath());
+      }
+    } catch (IOException e) {
+      LOG.error("Error while trying to delete expired proto events from " + eventsBasePath, e);
+    }
+  }
+
+  /**
+   * Delete the dir and if it fails, then retry deleting each files with file owner as proxy user.
+   */
+  private void deleteDir(FileSystem fs, FileStatus eventsDir) throws IOException {
+    try {
+      deleteByOwner(fs, eventsDir);
+      return;
+    } catch (IOException e) {
+      // Fall through.
+      LOG.info("Unable to delete the events dir " + eventsDir.getPath()
+              + " and so trying to delete event files one by one.");
+    }
+
+    FileStatus[] statuses = fs.listStatus(eventsDir.getPath());
+    for (FileStatus file : statuses) {
+      deleteByOwner(fs, file);
+    }
+    deleteByOwner(fs, eventsDir);
+  }
+
+  /**
+   * Delete the file/dir with owner as proxy user.
+   */
+  private void deleteByOwner(FileSystem fs, FileStatus fileStatus) throws IOException {
+    String owner = fileStatus.getOwner();
+    if (owner.equals(System.getProperty("user.name"))) {
+      fs.delete(fileStatus.getPath(), true);
+    } else {
+      LOG.info("Deleting " + fileStatus.getPath() + " as user " + owner);
+      UserGroupInformation ugi = UserGroupInformation.createProxyUser(owner,
+              UserGroupInformation.getLoginUser());
+      try {
+        ugi.doAs(new PrivilegedExceptionAction<Object>() {
+          @Override
+          public Object run() throws Exception {
+            fs.delete(fileStatus.getPath(), true);
+            return null;
+          }
+        });
+      } catch (InterruptedException ie) {
+        LOG.error("Could not delete " + fileStatus.getPath() + " for UGI: " + ugi, ie);
+      }
+      try {
+        FileSystem.closeAllForUGI(ugi);
```

Comment:
Some of these files can be owned by user hive, doing a closeAllForUGI closes every filesystem object owned by the user. It might be preferable to use the above approach, which will reduce the number of FileSystem instances and we can cleanup per directory.

Reply:
Fixed. Changed to new instance of file system and close it after deleting the dir.